### ISSUE TITLE
tests that call taskcluster services should be runnable locally for devs with credentials

### ISSUE THREAD

**Issue by petemoore** [taskcluster/taskcluster#3106]
To run the `taskcluster-notify` tests locally, I needed to fool the tests into thinking they were running in CI, by running `taskcluster-proxy` locally in one shell, `postgres` in another, and in a third shell, setting several following environment variables, before running `yarn workspace taskcluster-notify test`:

Shell 1:

```
$ docker run -ti -p 127.0.0.1:5432:5432 -e POSTGRES_HOST_AUTH_METHOD=trust --rm postgres:11 -c log_statement=all
```

Shell 2:

```
$ export TASKCLUSTER_CLIENT_ID=...
$ export TASKCLUSTER_ACCESS_TOKEN=...
$ export TASKCLUSTER_ROOT_URL=https://community-tc.services.mozilla.com
$ taskcluster-proxy
```

Shell 3:

```
$ export TASK_ID=<anything>
$ export TASKCLUSTER_PROXY_URL=http://localhost:8080
$ export NO_TEST_SKIP=true
$ export TEST_DB_URL=postgresql://postgres@localhost/postgres
$ export DEBUG=tc-lib-testing:*
$ yarn workspace taskcluster-notify test
```

 It would be neat, if tests would use `TASKCLUSTER_PROXY_URL` if and only if it is set, otherwise look for `TASKCLUSTER_{ACCESS_TOKEN,CLIENT_ID,CERTIFICATE}` and use those if they exist. We shouldn't need to check for presence of `TASK_ID` since tasks should be able to run both inside and outside of our CI automation. In other words, I think the automation should distinguish between the cases of taskcluster-proxy being present or not, rather than being "run in automation or not". This way, devs can choose whether to locally use `taskcluster-proxy` or not, and if a test really needed to / wanted to, it could also use real credentials.

In other words, I see `taskcluster-proxy` as being optional and transparent to the test, much like an http proxy can be used transparently to the application that uses it.

If we do this, the test won't care if it is running behind `taskcluster-proxy` or not, or whether it is running in automation or not, it will just care whether `NO_TEST_SKIP` is set or not, and the tests themselves will not be bound to our choice of CI environment (i.e. they could run equally well in e.g. jenkins/travis/appveyor etc where there might not be a `taskcluster-proxy`, but credentials might be set in e.g. encrypted environment variables).

**Comment by petemoore** [comment]
References:
* [Requirement TASK_ID present](https://github.com/taskcluster/taskcluster/blob/c838081727c15b345d7a8c6a29bc02da8fcbcbb7/libraries/testing/src/secrets.js#L21)
* [Requirement TASKCLUSTER_PROXY_URL present](https://github.com/taskcluster/taskcluster/blob/c838081727c15b345d7a8c6a29bc02da8fcbcbb7/libraries/testing/src/secrets.js#L83)

**Comment by djmitche** [comment]
Taskcluster-notify doesn't need Taskcluster credentials to run the tests.  It does need AWS credentials to test its email (SES) interface, and in CI it gets those AWS credentials from a TC secret.  But that's intended to occur only in CI, not locally.  Locally, you can just set the AWS_* variables if you want to test that functionality.  Looked at another way, it expects to find AWS credentials in environment variables, and there's a shim in place to set those environment variables from a secret when running in CI.

Both test files have
```
helper.secrets.mockSuite(testing.suiteName(), ['db', 'aws'], function(mock, skipping) {
```
meaning that they require both a db configuration and an aws configuration to run in "real" mode.  We could separate those by separating out the tests for the SES functionality (which require `aws` secrets, but don't use the DB) from the tests for all of the other service functionality (which would always use the fake SES, and use the real and fake DB in different runs of the suite).  That is,

```
# api_test.js, handler_test.js
...
helper.secrets.mockSuite(testing.suiteName(), ['db'], function(mock, skipping) {
  suiteSetup(function() {
    helper.load.inject('ses', new helper.MockSES());
  });
  ...
}

# ses_test.js
helper.secrets.mockSuite(testing.suiteName(), ['aws'], function(mock, skipping) {
  ...
}
```
In which case api_test.js and handler_test.js would run as long as TEST_DB_URL was set, but ses_test.js would not -- but for changes that don't modify the SES code, that's just fine.

Anyway, instead of making that refactor, the way to get your testing going right now is just to provide AWS credentials in environment variables.  Easiest will be to just copy these out of the existing TC secret and into your local shell.

Adding a section to the service's README giving the policy the tests require for that IAM user would be a good way to help the next person, or an external contributor with whom we do not want to share our AWS account credentials.  @imbstack might remember the details of that -- it may also require some SES pre-configuration?

**Comment by petemoore** [comment]
Thanks @djmitche 

It sounds like in this particular case, indeed I didn't need to jump through those hoops. However, in the general case, the same principle applies to any tests that require taskcluster credentials.

Currently, the test harness code that also runs outside of CI (`/libraries/testing/src/...`) _rather than the CI integration specific code (e.g. `.taskcluster.yml`)_ uses the presence of the `TASK_ID` environment variable to decide whether to connect to live taskcluster services, and then assumes that it should be done via taskcluster-proxy. I think independently of the taskcluster-notify tests, it would be beneficial for when `NO_TEST_SKIP` is explicitly set in any test, that the test harness should attempt to connect to a live taskcluster deployment, either via taskcluster-proxy if TASKCLUSTER_PROXY_URL is set, or via the regular TASKCLUSTER_* env vars if they are set, and TASKCLUSTER_PROXY_URL isn't set.

This way, _any_ tests that require taskcluster credentials can be run locally by devs, if they simply configure valid TASKCLUSTER_* credentials, or run a taskcluster-proxy if they really want to, and we should never need to limit tests to only being runnable in the CI. Note, the tests would continue to automatically run in CI by virtue of us setting NO_TEST_SKIP, but a developer would be able to do the same, should they wish to run one or more tests locally that require credentials to a taskcluster deployment.

I'll retitle the issue to better reflect its intended purpose...

**Comment by djmitche** [comment]
Let's talk face-to-face about this -- I think there's still a misunderstanding to be resolved.  Not urgent, though, so we can address it in a few weeks..

**Comment by djmitche** [comment]
I think we settled this back in June.
